lib/puppet/functions/partition.rb

```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: true
+
+# Returns two arrays, the first containing the elements of enum for which the block evaluates to true,
+# the second containing the rest.
+Puppet::Functions.create_function(:partition) do
+  # @param collection A collection of things to partition.
+  # @example Partition empty strings, results in e.g. [[''], [b, c]]
+  #   ['', b, c].group_by |$s| { $s.empty }
+  dispatch :partition do
+    required_param 'Iterable', :collection
+    block_param
+    return_type 'Tuple[Array, Array]'
+  end
+
+  def partition(collection, &block)
+    collection.partition(&block).freeze
```

Comment:
This will not work for all data types that are also Iterable (String for example).